### ISSUE TITLE
Coverage report refers to incorrect file URI (issue cross posted)

### ISSUE THREAD

**Issue by dulimarta** [t2ym/web-component-tester-istanbul#1]
I just posted [this issue](https://github.com/thedeeno/web-component-tester-istanbul/issues/42) at [the parent repo](https://github.com/thedeeno/web-component-tester-istanbul).

**Comment by t2ym** [comment]
@dulimarta I added a comment on [the original issue](https://github.com/thedeeno/web-component-tester-istanbul/issues/42).

**Comment by eugene-stepanov** [comment]
Hi @t2ym,

I experience the same issue. The problem is in the line of code below which constructs path by going a directory up having this '..' as a parameter to path.join() function:
https://github.com/t2ym/web-component-tester-istanbul/blob/e3da15b5105a5d1032e406f7fc0ea2bf7cf99420/lib/middleware.js#L29

Playing with the **--root** option doesn't help at all.

What is the reason for moving up a directory? Is it safe to remove '..' out of the function?

Thanks.

BTW: You've done a great job in making the whole thing working!

**Comment by t2ym** [comment]
Hi @eugene-stepanov,

Thank you for the comment.  I still cannot reproduce the issue.  In my environment, removing `'..'` does not solve the issue, only to show the `ENOENT` error instead.  The `'..'` is required because the trailing `/mywebapp` has to be removed from the root path to generate the joined path which is the absolute path to the file.

```sh
instrumentHtml: 
root = /home/user/wctissue/mywebapp
req.pathname = /mywebapp/src/my-view1.html 
htmlFilePath = /home/user/wctissue/mywebapp/src/my-view1.html
```

With the issue, the `htmlFilePath` seems to become `/home/user/wctissue/src/my-view1.html` unexpectedly and caused the `ENOENT` error.  So removing the `'..'` in your environment might solve the issue.

To identify the root cause, could you please tell me the following information?

- The current directory when invoking `wct`
- The command line options to `wct`
- The istanbul configuration in `wct.conf.json` or `wct.conf.js`
- The `ENOENT` error message in `my-view1.html.html`.
- The console output with the following patch to `node_modules/wct-istanbul/lib/middleware.js`.
```javascript
function instrumentHtml(html, req, root){
  var asset = req.url;
  var htmlFilePath = path.join(root, '..', parseurl(req).pathname);
  // Add the following line
  console.log('instrumentHtml: root = ' + root + ' req.pathname = ' + parseurl(req).pathname + ' htmlFilePath = ' + htmlFilePath);
```

Thanks in advance.

**Comment by eugene-stepanov** [comment]
Hi @t2ym,

Here is the info you've requested (_I'm on Windows system, sorry_ :wink:)
>- The current directory when invoking wct

`d:\projects\git\my-app`
>- The command line options to wct

No options just pure call: `wct`
> - The istanbul configuration in wct.conf.json or wct.conf.js:

```js
module.exports = {
	plugins: {
		local: {
			browsers: [ 'chrome' ],
		},
		istanbul: {
			dir: './coverage',
			reporters: [ 'text-summary', 'lcov' ],
			include: [ '\\build/inline/*.html' ],
		},
	},
	webserver: {
		port: 8081,
	},
	suites: [
		'build/test/*.test.html',
	],
};
```
> - The ENOENT error message in my-view1.html.html.

```
Error: ENOENT: no such file or directory, open 'd:\projects\git\build\inline\custom-component.html
```
There is no `my-app` in the file path. The expected path should be:
`d:\projects\git\my-app\build\inline\custom-component.html`

>- The console output with the following patch to node_modules/wct-istanbul/lib/middleware.js:

```
instrumentHtml:
    root = d:\projects\git\my-app
    req.pathname = /build/inline/custom-component.html
    htmlFilePath = d:\projects\git\build\inline\custom-component.html
```
The folders structure is:
```
+d:\projects\git\my-app
 \-- build
     \-- inline
     |   \-- custom-component.html
     |
     \-- test
         \-- custom-component.test.html
```

Thanks!

**Comment by t2ym** [comment]
Hi @eugene-stepanov, Thank you for the information. Let me look into the issue more deeply.

**Comment by t2ym** [comment]
## Fixes

- wct-istanbul@0.12.4
  - lib/middleware.js: https://github.com/t2ym/web-component-tester-istanbul/commit/19972e32f368d23fb58edad5c4939024e55c059c
- t2ym/polyserve#0.19.1-istanbul.3
  - package.json: https://github.com/t2ym/polyserve/commit/767aa4d2909b5438ac31c1d215ea1ec9f9aacd3e
  - TypeScript: https://github.com/t2ym/polyserve/commit/427160c1b1a8c8510ab589f955697f89e5202b12
  - Compiled JavaScript: https://github.com/t2ym/polyserve/commit/1ed2554174cad758541ddac26a32d34ae1f39178
- `@t2ym/web-component-tester` 6.0.2 : https://github.com/t2ym/web-component-tester/commit/3da9cd7010e1cc6d89c4a5f4959b5af5ce6cbe0d
  - `npm install --save-dev @t2ym/web-component-tester@^6.0.2`
  - `bower install --save-dev t2ym/web-component-tester#^6.0.2`

**Comment by t2ym** [comment]
Hi @eugene-stepanov,

Can you try these fixed modules?
- `wct-istanbul@0.12.4`
- `@t2ym/web-component-tester@^6.0.2`

## Root Cause:
- Requested URLs do not simply match with their paths in the file system
  - The path mapping scheme in polyserve/src/make_app.ts
```javascript
  app.get('*', (req, res) => {
    // Serve local files from . and other components from bower_components
    const url = parseUrl(req.url, true);
    let splitPath = url.pathname.split('/').slice(1);
    const splitPackagePath = packageName.split('/');
    if (arrayStartsWith(splitPath, splitPackagePath)) {
      if (root) {
        splitPath = [root].concat(splitPath.slice(splitPackagePath.length));
      } else {
        splitPath = splitPath.slice(splitPackagePath.length);
      }
    } else {
      splitPath = [componentDir].concat(splitPath);
    }
    const filePath = splitPath.join('/');
    req['_filePath'] = filePath; // Added for this issue
    const _send = send(req, filePath);
```

## Fix
- Put the mapped file path in the file system into `req._filePath` in `t2ym/polyserve#0.19.1-istanbul.3`
- Update `@t2ym/web-component-tester@6.0.2` to use `t2ym/polyserve#0.19.1-istanbul.3`
- Update `wct-istanbul@0.12.4` to use `t2ym/polyserve#0.19.1-istanbul.3`
- Pick up `req._filePath` if it exists
- Otherwise, remove the redundant component path
  - Redundant: root: `/home/user/my-element`, url path: `/my-element/my-element.html`
    - Note: The actual URL path is `/components/my-element/my-element.html`

**Comment by eugene-stepanov** [comment]
Hi @t2ym,

I checked out the updated modules. They now work just flawless and `LCOV` does it work as expected.

Thank you for quick fix, well done!

`BTW`: Is there a chance your forks be ever merged into original repos?

**Comment by t2ym** [comment]
Hi @dulimarta,

Is it possible for you to verify the fix with the following updated modules?

- `wct-istanbul@0.12.4`
- `@t2ym/web-component-tester@^6.0.2`

Thanks in advance.

**Comment by t2ym** [comment]
Hi @eugene-stepanov,

The merging is not a simple patch since my changes are full of tricks, which Google may not accept as a decent design, and `web-component-tester` will undergo major changes toward Polymer 3.0 with full NPM support.

The merging is tracked in the following issue.
https://github.com/t2ym/web-component-tester/issues/1

**Comment by t2ym** [comment]
The issue should be reproducible with `0.13.0` and `web-component-tester@6.6.0`

**Comment by kcmr** [comment]
I'm experiencing the same issue in a component.

The resulting path includes the scope and the package name after the root path instead of the filename.

The path should be `<root-path>/<filename>` but it is `<root-path>/<scope>/<package>/<filename>`.

This is the repo where I'm having this problem: https://github.com/kcmr/code-sample/tree/fix/coverage

**Comment by kcmr** [comment]
The following addition in `getFilePath()` (lib/middleware) solves the issue for me, but I'm not sure if this will cover other use cases:

```js
  var isScoped = /@/.test(componentPathName);

  if (isScoped) {
    var filename = pathname.split(/\//).pop();
    return req._filePath || path.join(root, filename);
  }
```

**Comment by t2ym** [comment]
Reopened for scope support

**Comment by t2ym** [comment]
@kcmr Can you try 0.14.3?  The same physical file path mapping logic as `polyserve` is copied from https://github.com/Polymer/tools/blob/master/packages/polyserve/src/make_app.ts

```javascript
    const url = parseUrl(req.url, true);
    let splitPath = url.pathname.split('/').slice(1);
    const splitPackagePath = packageName.split('/');

    if (arrayStartsWith(splitPath, splitPackagePath)) {
      if (root) {
        splitPath = [root].concat(splitPath.slice(splitPackagePath.length));
      } else {
        splitPath = splitPath.slice(splitPackagePath.length);
      }
    } else {
      splitPath = [componentDir].concat(splitPath);
    }
    const filePath = splitPath.join('/');
```

**Comment by kcmr** [comment]
@t2ym yes. It's solved in 0.14.3.
Thank you!
